### PR TITLE
Native prototype of ECMAScript `AsyncContext` proposal using existing public `v8::Context` methods

### PR DESCRIPTION
**Pull request by benjamn**
### tl;dr

These changes have been published in the [`benjamn/deno:async-context`](https://hub.docker.com/r/benjamn/deno/tags) Docker image, building on the work in PR #1, meaning you should be able to run my custom build of [Deno](https://deno.land/) (if you have [Docker](https://www.docker.com/) installed) using the following command:
```sh
docker run -it --rm benjamn/deno:async-context repl

# Alternatively, for 'thenables' support:
docker run -it --rm benjamn/deno:async-thenables repl
```

That command will open a JavaScript REPL where the [`AsyncContext`](https://github.com/legendecas/proposal-async-context/) constructor is available and fully implemented (according to my understanding of the proposal).

### Why now

I realize it's too early in the [TC39 proposal stage process](https://tc39.es/process-document/) to be deciding anything about the implementation. I am not advocating for this specific implementation (nor promoting Deno or Docker) over any other good ideas.

However, this is a subtle proposal, and many of the interested parties bring intuitions from previous systems with goals similar to those motivating `AsyncContext`, but with slightly different concrete usage, semantics, pitfalls, etc.

I hope this implementation allows those folks to check their intuitions thoroughly, especially in scenarios where the flow of `AsyncContext` could be needlessly lost (so explicit `AsyncContext.wrap` binding may be necessary).

Also, I believe everyone interested in these topics should at least be aware of [`v8::Context::GetContinuationPreservedEmbedderData` and `SetContinuationPreservedEmbedderData`](GetContinuationPreservedEmbedderData()), which provide a purpose-built way to hook into low-level `PromiseReactionJob` creation and execution, automatically bridging an all-important asynchronous gap in the ECMAScript event loop system.[^global-mutable-state]

[^global-mutable-state]: I don't think these `v8::Context::{Get,Set}ContinuationPreservedEmbedderData` methods should ever be exposed directly to userland code (global mutable state of the worst kind 😱), but I would argue `AsyncContext` wraps a safe abstraction over that fundamental capability, achieving continuity of `AsyncContext` across `PromiseReactionJob` creation and execution without exposing any of the necessary global state for tampering.

### Details

With this implementation, [`AsyncContext`](https://github.com/legendecas/proposal-async-context/) values are automatically preserved throughout the bodies of _native_ `async` functions, even after one or more `await` expressions, and also _not_ exposed during asynchronous pauses to other unrelated code that might happen to be running on the event loop in the meantime.

Similarly, typical use of `Promise`s (chaining `then` and `catch` and `finally`) should propagate `AsyncContext` as expected. As a bonus, `setTimeout` and `setInterval` now use `AsyncContext.wrap` to ensure context continuity for their callbacks.

As a concrete demonstration, these `expect` assertions pass:
```ts
const ctx = new AsyncContext;
ctx.run(1234, async () => {
  expect(ctx.get()).toBe(1234);
  const setTimeoutResult = await ctx.run(
    2345,
    () => new Promise(resolve => {
      setTimeout(() => resolve(ctx.get()), 20);
    }),
  );
  expect(setTimeoutResult).toBe(2345);
  expect(ctx.get()).toBe(1234);
  return "final result";
}).then(result => {
  expect(result).toBe("final result");
  // The code that generated the Promise has access to the 1234
  // value provided to ctx.run above, but consumers of the Promise
  // do not automatically inherit it.
  expect(ctx.get()).toBeUndefined();
});
```
If you want something that's easier to paste into the Deno REPL, try this:
```ts
const ctx = new AsyncContext;
ctx.run(1234, async () => {
  console.log("before await", ctx.get());
  const setTimeoutResult = await ctx.run(
    2345,
    () => new Promise(resolve => {
      setTimeout(() => resolve(ctx.get()), 20);
    }),
  );
  console.log("after await", ctx.get());
  console.log({ setTimeoutResult });
  return "final result";
}).then(result => {
  console.log("final result?", result);
  console.log("outside ctx.run", ctx.get());
})
```
If all goes well, the console output should be:
```
before await 1234
Promise { <pending> }
> after await 1234
{ setTimeoutResult: 2345 }
final result? final result
outside ctx.run undefined
```

### Why did I choose Deno?

A few reasons, none of which conflict with my love of Node.js:
* I wanted to improve my Rust skills
  * Please let me know if you spot any better ways to do the Rust parts
* You could probably implement the `AsyncContext` proposal using the Node.js [`async_hooks`](https://nodejs.org/api/async_context.html) utilities (like [`AsyncLocalStorage`](https://nodejs.org/api/async_context.html#class-asynclocalstorage)), but I wanted to show that is not necessary
  * V8 already provides the raw materials via [two existing methods of `v8::Context`](https://github.com/v8/v8/blob/10.9.194.4/src/api/api.cc#L6709-L6725), which were introduced in [this cryptic/convenient V8 pull request](https://chromium-review.googlesource.com/c/v8/v8/+/2005849) back in March 2020[^aside]
* Deno builds to a single standalone static binary, which makes it possible to distribute with a relatively small Docker image (162MB total, 89MB of which are that `deno` binary)

[^aside]: World events may have overshadowed the news of this pull request landing in V8

I'd be happy to try something similar with Node.js, if there's interest.

### Why am I using Docker?

If you happen to have Docker installed, downloading and running a ready-to-go 162MB image is likely to be much faster than cloning the necessary repositories and submodules and building everything from source, which can take more than an hour on my late-model Intel Macbook Pro.

If you don't have access to Docker, the [`docker/`](https://github.com/benjamn/deno/tree/main/docker) directory is still a good reference for how to build the project from source.

Building from source may also be worthwhile if you want to use the custom version of `deno` as your `deno.path` in `.vscode/settings.json`, so that (for example) `AsyncContext` will be known by the type system to be globally defined when using the [Deno VSCode integration](https://deno.land/manual@v1.29.4/references/vscode_deno).

### CODE REVIEW

File: docker/tests/adapted-from-proposal.ts [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+type Value = { id: number };
+function sleep(ms: number): Promise<void> {
+  return new Promise((r) => setTimeout(r, ms));
+}
+
+describe("sync", () => {
+  describe("run and get", () => {
+    it("has initial undefined state", () => {
+      const ctx = new AsyncContext<Value>();
```
Comment (benjamn) [inline docker/tests/adapted-from-proposal.ts:15]: These (minimally adapted) tests that I borrowed from [the proposal repo](https://github.com/legendecas/proposal-async-context/blob/master/tests/async-context.test.ts) are [passing](https://github.com/benjamn/deno/actions/runs/3976580483/jobs/6817158512), without any `Promise.prototype.then` wrapping. Pretty cool!

I will add some stronger tests in a subsequent commit/PR.

File: docker/tests/adapted-from-proposal.ts [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+describe("async via promises", () => {
+  // This wrapping is no longer needed, since the native implementation of
+  // AsyncContext hooks into native PromiseReactionJob creation and execution.
+  //
+  // beforeEach(() => {
+  //   Promise.prototype.then = then;
+  // });
+  // afterEach(() => {
+  //   Promise.prototype.then = nativeThen;
+  // });
```
Comment (benjamn) [inline docker/tests/adapted-from-proposal.ts:144]: This is what I meant by avoiding "`Promise.prototype.then` wrapping."

File: docker/tests/additional-native-tests.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    const ctxRunResult = await ctx.run(1234, async () => {
+      assert.strictEqual(ctx.get(), 1234);
+      const setTimeoutResult = await ctx.run(
+        2345,
+        () => new Promise(resolve => {
+          setTimeout(() => resolve(ctx.get()), 20);
+        }),
+      );
+      assert.strictEqual(setTimeoutResult, 2345);
+      assert.strictEqual(ctx.get(), 1234);
```
Comment (benjamn) [inline docker/tests/additional-native-tests.ts:21]: Here are some native `async`/`await` tests that also demonstrate the `setTimeout` integration.

**Comment by benjamn** [comment]
That failing test was using the old `benjamn/deno:async-context` image, which did not fully support `AsyncContext` with "thenable" objects (any object with a `then` method). Once I update the test to use the newer `benjamn/deno:async-thenables` Docker image, the new tests should all pass…

File: docker/tests/additional-native-tests.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      assert.strictEqual(thenableResult, "thenable: running");
+      assert.strictEqual(ctx.get(), void 0);
+      assert.deepStrictEqual(queue, [
+        "thenable: running",
+        "thenable: running",
+        "thenable: inner",
+        "thenable: running",
+      ]);
+    });
+  });
```
Comment (benjamn) [inline docker/tests/additional-native-tests.ts:80]: This new test demonstrates `AsyncContext` is now preserved even when using `Promise`-like _thenable_ objects (any object with a function-valued `then` method). This behavior did not work with my previous Docker image `benjamn/deno:async-context`, but it works now with `benjamn/deno:async-thenables`.

Technically, adding support for thenables boiled down to making [`PromiseResolveThenableJobTask`](https://github.com/benjamn/deno-v8/blob/d72a3dd59139242385203b3aa0167451220d0bdf/src/builtins/builtins-microtask-queue-gen.cc#L209-L232) behave more like [`PromiseResolveReactionJob`](https://github.com/benjamn/deno-v8/blob/d72a3dd59139242385203b3aa0167451220d0bdf/src/builtins/builtins-microtask-queue-gen.cc#L257-L288), in https://github.com/benjamn/deno-v8/commit/d72a3dd59139242385203b3aa0167451220d0bdf, my fork of Deno's fork of V8.

This change was inspired by two comments from @jridgewell:
* https://github.com/nodejs/node/issues/46265#issuecomment-1397496044
* https://github.com/nodejs/node/issues/46265#issuecomment-1397764624

There might be a better way to handle all kinds of host reaction jobs at once, but improving `AsyncContext` support for thenables to match other patterns of `Promise` and `async`/`await` usage seems like an improvement to me.

I know it's not this simple to change how V8 processes thenables/anything, with so many dependent projects (Node.js, Deno, Chromium, etc) who would all have to be on board (or indifferent). I also know I'm probably not showing the people who make these decisions anything they don't already know, but for me it's been useful to play with a prototype where `AsyncContext` and thenables work together.

In other words, I don't claim to be fully addressing the last remaining concern expressed in https://github.com/cloudflare/workerd/pull/282#issuecomment-1402921124, but here's a REPL you can play with right now to see what I'm talking about:
```sh
docker run -it --rm benjamn/deno:async-thenables repl
```
Try pasting in this code:
```ts
const c = new AsyncContext();
c.run(1234, async () => {
  console.log("before await", c.get());
  const onResResult = await {
    then(onRes) {
      return Promise.resolve(onRes(c.get()));
    }
  };
  console.log("after await", c.get(), { onResResult });
})
```
If all goes well, that code should log
```
before await 1234
after await 1234 { onResResult: 1234 }
Promise { undefined }
```
However, if you try the same code with my previous Docker image `benjamn/deno:async-context`…
```sh
docker run -it --rm benjamn/deno:async-context repl
```
then the result will be
```
before await 1234
after await 1234 { onResResult: undefined }
Promise { undefined }
```
For more examples, see the test code I'm commenting on.

Comment (jridgewell) [reply docker/tests/additional-native-tests.ts:80]: Lol, we were just talking about making these exact changes yesterday on our Matrix chat. If you're on Matrix, I'd love to add you to our group chat.

> There might be a better way to handle all kinds of host reaction jobs at once, but improving AsyncContext support for thenables to match other patterns of Promise and async/await usage seems like an improvement to me.

Yes, I think instead of implementing these for each individual enqueable task type, the queue itself should implement this for all types. But for now, this fix is exactly what I recommended in our chat.

Comment (benjamn) [reply docker/tests/additional-native-tests.ts:80]: > If you're on Matrix, I'd love to add you to our group chat.

Sure! I'm a Matrix newb though I've used it a bit for TC39. I think `@benjamn:matrix.org` is my handle?

Comment (jridgewell) [reply docker/tests/additional-native-tests.ts:80]: You're invited.

Comment (Jbnado) [reply docker/tests/additional-native-tests.ts:80]: What is this Matrix?

Comment (jridgewell) [reply docker/tests/additional-native-tests.ts:80]: https://matrix.to/#/#tc39-async-context:matrix.org
